### PR TITLE
BREAKING - NLS: Adds a method which converts a given time range in milliseconds to a time representation format typically found in digital clocks

### PR DESCRIPTION
**Pull request by mko-sci**
Fixes: [OX-9555](https://scireum.myjetbrains.com/youtrack/issue/OX-9555/Video-Modernisierung-Offene-Punkte), [SIRI-731](https://scireum.myjetbrains.com/youtrack/issue/SIRI-731/NLS-convertDuration-refactorn)

### CODE REVIEW

File: src/main/java/sirius/kernel/nls/NLS.java [hunk 1]
```diff
@@ -1315,6 +1315,20 @@ private static void appendDurationValue(StringBuilder result, String key, long v
         result.append(get(key, (int) value));
     }
 
+    /**
+     * Converts the given time range in milliseconds to a time representation format typically found in digital clocks.
```
Comment (sabieber) [inline src/main/java/sirius/kernel/nls/NLS.java:1319]:
```suggestion
     * Converts the given time range in milliseconds to a time representation format typically found in digital clocks (like 12:05:17).
```

Wär schon nett wenn das JavaDoc sagt wie es hinterher aussieht :)

**Review by sabieber — APPROVED** [review]

File: src/main/java/sirius/kernel/nls/NLS.java [hunk 2]
```diff
@@ -1273,59 +1273,97 @@ public static <V> V parseUserString(Class<V> clazz, String string) {
...
+            appendDurationValue(result, "NLS.minutes", minutes);
+        }
+        int seconds = duration.toSecondsPart();
+        if (includeSeconds && seconds > 0 || includeMillis) {
+                appendDurationValue(result, "NLS.seconds", seconds);
+        }
+        int millis = duration.toMillisPart();
+        if (includeMillis && millis > 0) {
+            appendDurationValue(result, "NLS.milliseconds", millis);
         }
```
Comment (sabieber) [inline src/main/java/sirius/kernel/nls/NLS.java:1306]: Syntactically this is not entirely the same logic as before.
Before: includeMillis only was applied when includeSeconds AND includeMillis is set to true
After: includeMillis can be set to true while includeSeconds is disabled

File: src/main/java/sirius/kernel/nls/NLS.java [hunk 3]
```diff
@@ -1273,59 +1273,97 @@ public static <V> V parseUserString(Class<V> clazz, String string) {
...
+     *
+     * @param duration       the duration in milliseconds
+     * @param includeSeconds determines whether to include seconds or to ignore everything below minutes
+     * @param includeMillis  determines whether to include milliseconds or to ignore everything below seconds
+     * @return a string representation of the given duration in days, hours, minutes and,
+     * if enabled, seconds and milliseconds
+     */
+    @Deprecated(forRemoval = true)
+    public static String convertDuration(long duration, boolean includeSeconds, boolean includeMillis) {
+        return convertDuration(Duration.ofMinutes(duration), includeSeconds, includeMillis);
```
Comment (sabieber) [inline src/main/java/sirius/kernel/nls/NLS.java:1335]:
```suggestion
        return convertDuration(Duration.ofMillis(duration), includeSeconds, includeMillis);
```

**Review by sabieber — CHANGES_REQUESTED** [review]

File: src/main/java/sirius/kernel/nls/NLS.java [hunk 4]
```diff
@@ -1273,69 +1273,97 @@ public static <V> V parseUserString(Class<V> clazz, String string) {
...
+        long days = duration.toDaysPart();
+        if (days > 0) {
+            appendDurationValue(result, "NLS.days", days);
+        }
+        int hours = duration.toHoursPart();
+        if (hours > 0) {
+            appendDurationValue(result, "NLS.hours", hours);
+        }
+        int minutes = duration.toMinutesPart();
+        if (minutes > 0) {
```
Comment (idlira) [inline src/main/java/sirius/kernel/nls/NLS.java:1296]: Exact minutes will not be added? like exact 3 hours to be 03:00... ?

Comment (idlira) [reply src/main/java/sirius/kernel/nls/NLS.java:1296]: A unit test would be nice to verify diverse variants.

**Review by idlira — CHANGES_REQUESTED** [review]

File: src/test/java/sirius/kernel/nls/NLSSpec.groovy [hunk 5]
```diff
@@ -333,4 +333,125 @@ class NLSSpec extends BaseSpecification {
...
+        NLS.convertDurationToDigitalClockFormat(input) == output
+        where:
+        input                                               | output
+        Duration.ofMinutes(60L)                              | "01:00:00"
+        Duration.ofMinutes(60L).plusSeconds(1L)              | "01:00:01"
+        Duration.ofHours(1).plusMinutes(1L)                 | "01:01:00"
+        Duration.ofHours(1).plusMinutes(1L).plusSeconds(1L) | "01:01:01"
+        Duration.ofDays(2l)                                 | "48:00:00"
+        Duration.ofDays(2l).plusHours(1L)                   | "49:00:00"
+        Duration.ofDays(2l).plusMinutes(1L)                 | "48:01:00"
```
Comment (sabieber) [inline src/test/java/sirius/kernel/nls/NLSSpec.groovy:348]:
```suggestion
        Duration.ofDays(2L)                                 | "48:00:00"
        Duration.ofDays(2L).plusHours(1L)                   | "49:00:00"
        Duration.ofDays(2L).plusMinutes(1L)                 | "48:01:00"
```

**Review by sabieber — APPROVED** [review]

**Review by idlira — APPROVED** [review]

File: src/test/java/sirius/kernel/nls/NLSSpec.groovy [hunk 6]
```diff
@@ -333,4 +333,125 @@ class NLSSpec extends BaseSpecification {
...
+    }
+
+    def "convertDuration() of Duration is properly formatted"() {
+        expect:
+        NLS.convertDuration(duration, enableSeconds, enableMillis) == output
+        where:
+        duration                  | enableSeconds | enableMillis |
+                output
+        Duration
+                .ofMinutes(60L)   | true          | true         | "1 Stunde"
```
Comment (fhaScireum) [inline src/test/java/sirius/kernel/nls/NLSSpec.groovy:361]: Ist das richtig formattiert?

**Review by fhaScireum — APPROVED** [review]

File: src/test/java/sirius/kernel/nls/NLSSpec.groovy (hunk 6 above)
Comment (mko-sci) [reply src/test/java/sirius/kernel/nls/NLSSpec.groovy:361]: nö, aber jetzt. 
Da kackt IntelliJ regelmäßig ab...

**Review by fhaScireum — APPROVED** [review]

File: src/test/java/sirius/kernel/nls/NLSSpec.groovy (hunk 6 above)
Comment (sabieber) [reply src/test/java/sirius/kernel/nls/NLSSpec.groovy]: Ein Grund wieso Groovy weg soll :)

**Review by andyHa — APPROVED** [review]

**Pull request merged**